diff --git a/node_modules/regodit/regedit.js b/node_modules/regodit/regedit.js
index e803f5d..400cf4a 100644
--- a/node_modules/regodit/regedit.js
+++ b/node_modules/regodit/regedit.js
@@ -3,7 +3,7 @@
 const path = require('path');
 const ffi = require('ffi-napi');
 
-const lib = ffi.Library(path.join(__dirname, "build/regedit.x64.dll"), {
+const lib = ffi.Library(path.join(__dirname, "build/regedit.x64.dll").replace('app.asar', 'app.asar.unpacked'), {
    'RegKeyExists': ["int", ["string", "string"]],
    'RegListAllSubkeys': ["string", ["string", "string"]],
    'RegListAllValues': ["string", ["string", "string"]],
