diff --git a/node_modules/regodit/lib/regedit.cjs b/node_modules/regodit/lib/regedit.cjs
index 8a6fc11..ab82035 100644
--- a/node_modules/regodit/lib/regedit.cjs
+++ b/node_modules/regodit/lib/regedit.cjs
@@ -4,7 +4,7 @@ const path = require('path');
 const ffi = require('ffi-napi');
 const normalize = require('./normalize.cjs');
 
-const lib = ffi.Library(path.join(__dirname, "dist/regodit.x64.dll"), {
+const lib = ffi.Library(path.join(__dirname, "dist/regodit.x64.dll").replace('app.asar', 'app.asar.unpacked'), {
    'RegKeyExists': ["int", ["string", "string"]],
    'RegListAllSubkeys': ["string", ["string", "string"]],
    'RegListAllValues': ["string", ["string", "string"]],
