diff --git a/node_modules/regodit/lib/regedit.cjs b/node_modules/regodit/lib/regedit.cjs
index fe10d60..fbf44e5 100644
--- a/node_modules/regodit/lib/regedit.cjs
+++ b/node_modules/regodit/lib/regedit.cjs
@@ -6,7 +6,7 @@ const normalize = require('./util/normalize.cjs');
 
 const HKEY = require("./HKEY.json");
 
-const lib = ffi.Library(path.join(__dirname, "dist/regodit.dll"), {
+const lib = ffi.Library(path.join(__dirname, "dist/regodit.dll").replace('app.asar', 'app.asar.unpacked'), {
    'RegKeyExists': ["int", ["string", "string"]],
    'RegListAllSubkeys': ["string", ["string", "string"]],
    'RegListAllValues': ["string", ["string", "string"]],
