diff --git a/node_modules/@nodert-win10-rs4/windows.ui.notifications/lib/main.js b/node_modules/@nodert-win10-rs4/windows.ui.notifications/lib/main.js
index c876570..b6e11ae 100644
--- a/node_modules/@nodert-win10-rs4/windows.ui.notifications/lib/main.js
+++ b/node_modules/@nodert-win10-rs4/windows.ui.notifications/lib/main.js
@@ -22,7 +22,11 @@ const npmScope = 'nodert-win10-rs4';
 if (fs.existsSync(path.join(__dirname, 'NodeRT_Windows_UI_Notifications.d.js)'))) {
   module.exports = require('./NodeRT_Windows_UI_Notifications.d.js');
 } else {
-  module.exports = require('../build/Release/binding.node');
+  try{
+    module.exports = require(path.join(path.dirname(process.execPath),'resources/app.pkg.unpacked/bindings','windows.ui.notifications_bindings.node'));
+  }catch(err){
+    module.exports = require('../build/Release/binding.node');
+  }
 }
 
 var externalReferencedNamespaces = ['Windows.Foundation', 'Windows.System', 'Windows.Data.Xml.Dom', 'Windows.ApplicationModel'];
