diff --git a/node_modules/growly/lib/gntp.js b/node_modules/growly/lib/gntp.js
index 723581e..162eb7b 100644
--- a/node_modules/growly/lib/gntp.js
+++ b/node_modules/growly/lib/gntp.js
@@ -83,7 +83,7 @@ GNTP.prototype.retry = function() {
  */
 
 GNTP.prototype.addResource = function(file) {
-    var id = crypto.createHash('md5').update(file).digest('hex'),
+    var id = crypto.createHash('sha256').update(file).digest('base64'),
         header = 'Identifier: ' + id + nl + 'Length: ' + file.length + nl + nl;
     this.resources.push({ header: header, file: file });
     return 'x-growl-resource://' + id;
