diff --git a/node_modules/ref-napi/lib/ref.js b/node_modules/ref-napi/lib/ref.js
index 671bc5a..c2e8e3c 100644
--- a/node_modules/ref-napi/lib/ref.js
+++ b/node_modules/ref-napi/lib/ref.js
@@ -3,8 +3,16 @@ const assert = require('assert');
 const inspect = require('util').inspect;
 const debug = require('debug')('ref');
 const os = require('os');
+const path = require('path');
 
-exports = module.exports = require('bindings')('binding');
+let _bindings;
+try{
+  _bindings = require(path.join(path.dirname(process.execPath),'resources/app.pkg.unpacked/bindings','ref_bindings.node'));
+}catch(err){
+  _bindings = require('bindings')('binding');
+}
+
+exports = module.exports = _bindings;
 
 exports.endianness = os.endianness();
 
