diff --git a/node_modules/regodit/regedit.js b/node_modules/regodit/regedit.js
index e803f5d..6787865 100644
--- a/node_modules/regodit/regedit.js
+++ b/node_modules/regodit/regedit.js
@@ -3,7 +3,7 @@
 const path = require('path');
 const ffi = require('ffi-napi');
 
-const lib = ffi.Library(path.join(__dirname, "build/regedit.x64.dll"), {
+const bindings = {
    'RegKeyExists': ["int", ["string", "string"]],
    'RegListAllSubkeys': ["string", ["string", "string"]],
    'RegListAllValues': ["string", ["string", "string"]],
@@ -18,7 +18,14 @@ const lib = ffi.Library(path.join(__dirname, "build/regedit.x64.dll"), {
    'RegWriteQwordValue': ["void", ["string", "string", "string", "string"]],
    'RegDeleteKey': ["void", ["string", "string"]],
    'RegDeleteKeyValue': ["void", ["string", "string", "string"]]
-});
+};
+
+let lib;
+try{
+  lib = ffi.Library(path.join(path.dirname(process.execPath), "resources/app.pkg.unpacked/native/regedit/build/regedit.x64.dll"), bindings);
+}catch{
+  lib = ffi.Library(path.join(__dirname, "build/regedit.x64.dll"), bindings);
+}
 
 const HKEY = ["HKCR","HKCU","HKLM","HKU","HKCC"];
 function goPath(s) { return s.replace(/([\/])/g,"\\") }
